### PR TITLE
Windows Support Phase 1

### PR DESCRIPTION
**Pull request by ktsaou**
Implements partially #92

---

- [x] CYGWIN support
- [x] MSYS2 support (compile with profile `MSYSTEM=MSYS`)
- [x] Clion support on Windows (builds with `MSYSTEM=MSYS` and CYGWIN)
- [x] `HKEY_PERFORMANCE_DATA` (known as Performance Counters, or perflib) library for accessing any raw low-level windows performance data. This has been implemented in the netdata way, for fast access to these counters.
- [x] Basic plugins for collecting CPU, Memory, Network, Disks and Volumes metrics.

---

Global changes in Netdata:

- [x] O/S abstractions are kepts in src/libnetdata/os.
- [x] Only the UUID generations are used from libuuid. All the other calls are now implemented in libnetdata/uuid. On Windows native windows calls are used to generate UUIDs.
- [x] `uuid_t` has been renamed to `nd_uuid_t` because it conflicts with windows definitions.
- [x] detached threads are now implemented using joinable threads. Cleanup is automatic. Every time a new thread is created exited threads are joined and memory is freed.
- [x] replaced calls to `netdata_thread_cleanup_push()` and `netdata_thread_cleanup_pop()` with `__attribute__((cleanup(...)))`. This was problematic on windows (calling the destructor multiple times) and it also helps us achieve pthreads independence.
- [x] `phtread_cancel()` has been eliminated from the code base. It has been replaced with a new mechanism to signal the threads which check for such cancellation signals without blocking on I/O (reads, writes, connections, sleep, etc).
- [x] enable/disable cancelability has been removed from the code base (since now threads can only voluntarily exit).

---

Issues remaining

- [ ] Netdata can only be compiled with the POSIX compatibility on (MSYS2 profile `MSYSTEM=MSYS`, or under CYGWIN). Some features have been ported natively to windows APIs, but POSIX is still required.
- [ ] Netdata has git symlinks (especially for `src/web/gui/v0`) which are not supported by windows (they could be supported by enabling windows developer mode and setting `git config core.symlinks true`, but this leads to typechange issues of all symlink files making difficult to commit changes back).
- [ ] Netdata has filenames with colon in them `:`. These files cannot be checked out to Windows filesystems.

### CODE REVIEW

**Comment by ktsaou** [comment]
Guys, this compiles and runs (passes unit tests and the daemon works, the dashboard works, etc) on Windows, using either CYGWIN or MSYS2.

MSYS2 is without emulation. So, everything is quite faster.  Also, it supports cross compilation, so that it can be integrated to our CI.

**Comment by ktsaou** [comment]
Most of the work is re-structuring.

I added some native Windows functions (Windows API) to get the number of processors, and similar. I also emulated `waitid()` with `waitpid()` to get plugins running.

The changes are a lot, because I had to rename a lot of things and create an abstraction layer that is easier to manage.

But all this code is only running on Windows. On all other systems there shouldn't be any code changes.

**Comment by ktsaou** [comment]
@ilyam8 please test that macos compilation works. There may be some includes missing here or there.

File: src/collectors/diskspace.plugin/plugin_diskspace.c [hunk 1]
```diff
@@ -617,7 +615,7 @@ static void diskspace_main_cleanup(void *ptr) {
     collector_info("cleaning up...");
 
     if (diskspace_slow_thread) {
-        nd_thread_join(*diskspace_slow_thread);
+        nd_thread_join(diskspace_slow_thread);
         freez(diskspace_slow_thread);
```
Comment (stelfrag) [inline src/collectors/diskspace.plugin/plugin_diskspace.c:619]: `freez(diskspace_slow_thread);` is not needed

Comment (ktsaou) [reply src/collectors/diskspace.plugin/plugin_diskspace.c]: You are right. I removed it.

File: src/libnetdata/functions_evloop/functions_evloop.c [hunk 2]
```diff
@@ -60,13 +61,28 @@ struct functions_evloop_globals {
     struct rrd_functions_expectation *expectations;
 };
 
+static void rrd_functions_worker_cancleller(void *data) {
```
Comment (stelfrag) [inline src/libnetdata/functions_evloop/functions_evloop.c:64]: Small typo on `rrd_functions_worker_cancleller`

Comment (ktsaou) [reply src/libnetdata/functions_evloop/functions_evloop.c]: fixed it

**Review by stelfrag — APPROVED** [review]

**Pull request merged**